### PR TITLE
Add subregion concept to cluster mode

### PR DESCRIPTION
**Pull request by donggyu04**
Currently, if you run ngrinder in cluster mode, you can configure only one region for each controllers.
To improve this inefficiency, I allow to configure subregions under original region.

#### 1. How to configure the subregion.

You can configure the subregion by modifying the `system-ex.conf` file like below. Subregion only active in cluster mode.

```
...
ngrinder.cluster.region=korea
ngrinder.cluster.subregion=sub1,sub2
...
```

#### 2. Feature

1.  Agents can be divided and managed from one controller through subregion.

### CODE REVIEW

File: ngrinder-controller/src/main/java/org/ngrinder/agent/controller/AgentDownloadController.java [hunk 1]
```diff
@@ -73,11 +74,20 @@ public void download(@PathVariable String fileName, HttpServletResponse response
 	 * @param request request.
 	 */
 	@GetMapping("/{region}/{owner}")
-	public String downloadDirect(@PathVariable(value = "owner") String owner,
-	                             @PathVariable(value = "region") String region,
+	public String downloadDirect(@PathVariable String owner,
+	                             @PathVariable String region,
```
Comment (junoyoon) [inline ngrinder-controller/src/main/java/org/ngrinder/agent/controller/AgentDownloadController.java:78]: switch the param order to follow their apearance.

File: ngrinder-controller/src/main/java/org/ngrinder/common/util/RegionUtils.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	/**
+	 * Convert configured string type of subregion value to {@link List}.
+	 *
+	 * @param subregionsString value of 'ngrinder.cluster.subregion' in system-ex.conf, it concatenates with ',' like sub1,sub2
+	 */
+	public static List<String> convertSubregionsStringToList(String subregionsString) {
+		return isEmpty(subregionsString) ? emptyList() : asList(subregionsString.split(","));
+	}
+
+}
```
Comment (junoyoon) [inline ngrinder-controller/src/main/java/org/ngrinder/common/util/RegionUtils.java:56]: Is it valuable util to exist separately?

File: ngrinder-controller/src/main/java/org/ngrinder/infra/config/Config.java [hunk 3]
```diff
@@ -207,6 +212,22 @@ public String getRegion() {
 		return isClustered() ? getClusterProperties().getProperty(PROP_CLUSTER_REGION) : NONE_REGION;
 	}
 
+	/**
+	 * Get the current region and subregion from the configuration.
+	 *
+	 * @return region and subregion. If it's not clustered mode, return "NONE"
+	 */
+	public Map<String, String> getRegionWithSubregion() {
```
Comment (junoyoon) [inline ngrinder-controller/src/main/java/org/ngrinder/infra/config/Config.java:220]: Why do you return map here?

File: ngrinder-controller/src/main/java/org/ngrinder/perftest/repository/PerfTestRepository.java [hunk 4]
```diff
@@ -60,7 +60,7 @@
 	 * @param region region where the test belong to
 	 * @return perf test list
 	 */
-	List<PerfTest> findAllByStatusAndRegionOrderByScheduledTimeAsc(Status status, String region);
+	List<PerfTest> findAllByStatusAndRegionStartsWithOrderByScheduledTimeAsc(Status status, String region);
```
Comment (junoyoon) [inline ngrinder-controller/src/main/java/org/ngrinder/perftest/repository/PerfTestRepository.java:63]: StartsWithOrder??

**Comment by donggyu04** [comment]
Use dropdown component to select region. 

#### 1. Perftest configuration


#### 2. Agent management


#### 3. Agent download

File: ngrinder-controller/src/main/java/org/ngrinder/infra/config/Config.java (hunk 3 above)
Comment (donggyu04) [reply ngrinder-controller/src/main/java/org/ngrinder/infra/config/Config.java:220]: I use [this method](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/config/MemberAttributeConfig.java#L42) to add attributes to hazelcast member.
It is used when clustered hazelcast members share own region information.

Comment (donggyu04) [reply ngrinder-controller/src/main/java/org/ngrinder/infra/config/Config.java:220]: Hazelcast member attribute is accept only String type.

```
// Map data format

"region": "kor"
"subregion: "sub1,sub2,sub3"
```

File: ngrinder-controller/src/main/java/org/ngrinder/perftest/repository/PerfTestRepository.java (hunk 4 above)
Comment (donggyu04) [reply ngrinder-controller/src/main/java/org/ngrinder/perftest/repository/PerfTestRepository.java:63]: After add subregion. region is stored to the database like `{region}.{subregion}`.
So I use StartsWith to select all of the perftests that have configured region or subregion from the database.

File: ngrinder-controller/src/main/java/org/ngrinder/common/util/RegionUtils.java (hunk 2 above)
Comment (donggyu04) [reply ngrinder-controller/src/main/java/org/ngrinder/common/util/RegionUtils.java:56]: Convert to list is unnecessary, I will remove it.

File: ngrinder-controller/src/main/java/org/ngrinder/agent/controller/AgentDownloadController.java (hunk 1 above)
Comment (donggyu04) [reply ngrinder-controller/src/main/java/org/ngrinder/agent/controller/AgentDownloadController.java]: Fixed all.

**Review by junoyoon — APPROVED** [review]

File: ngrinder-core/src/main/java/net/grinder/AgentController.java [hunk 5]
```diff
@@ -95,12 +96,29 @@ public AgentController(Condition eventSyncCondition, AgentConfig agentConfig) {
...
-		this.m_agentIdentity.setRegion(agentConfig.getRegion());
+		this.m_agentIdentity.setSubregion(agentConfig.getSubregion());
+
+		String legacyOwner = getOwnerFromLegacyRegionConfig(agentConfig.getRegion());
+		String configuredOwner = agentConfig.getOwner();
+
+		m_agentIdentity.setOwner(legacyOwner);
+		if (isNotEmpty(configuredOwner)) {
+			this.m_agentIdentity.setOwner(configuredOwner);
+		}
```
Comment (imbyungjun) [inline ngrinder-core/src/main/java/net/grinder/AgentController.java:107]: what about using `if - else`? and add `this` to first statement?

Comment (donggyu04) [reply ngrinder-core/src/main/java/net/grinder/AgentController.java]: It is Fixed

**Review by junoyoon — APPROVED** [review]

**Review by songeunwoo — APPROVED** [review]

**Pull request merged**